backend/controller/controller.go

```diff
@@ -693,16 +695,53 @@ func (s *Service) GetModuleContext(ctx context.Context, req *connect.Request[ftl
 			return connect.NewError(connect.CodeInternal, fmt.Errorf("could not get databases: %w", err))
 		}
 
-		response := modulecontext.NewBuilder(name).AddConfigs(configs).AddSecrets(secrets).AddDatabases(databases).Build().ToProto()
+		checksum := configurationMapChecksum(configs)
+		checksum = (checksum * 115163) + configurationMapChecksum(secrets)
+		checksum = (checksum * 454213) + configurationDatabaseChecksum(databases)
+
+		if checksum != lastChecksum {
+			response := modulecontext.NewBuilder(name).AddConfigs(configs).AddSecrets(secrets).AddDatabases(databases).Build().ToProto()
+
+			if err := resp.Send(response); err != nil {
+				return connect.NewError(connect.CodeInternal, fmt.Errorf("could not send response: %w", err))
+			}
 
-		if err := resp.Send(response); err != nil {
-			return connect.NewError(connect.CodeInternal, fmt.Errorf("could not send response: %w", err))
+			lastChecksum = checksum
 		}
 
-		time.Sleep(s.config.ModuleUpdateFrequency)
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-time.After(s.config.ModuleUpdateFrequency):
+		}
 	}
 }
 
+// configurationMapChecksum computes a checksum on the map that is order invariant.
+//
+// This operation is used to detect configuration change.
+func configurationMapChecksum(m map[string][]byte) int64 {
+	sum := int64(0)
+	for k, v := range m {
```

Comment:
Ranging over a map is not ordered, you'll need to use `maps.Keys(m)` and `sort.Slice()`.

Reply:
Will make these adjustments in the follow-up PR. I went with a sum of hashes to avoid having to worry about ordering; after adopting the suggestions the randomized ordering will matter

Reply:
Ah yeah I realised after that you were adding the values to avoid the sort, but I think just using sha256 directly will be clearer and simpler.